File: app/src/main/java/nl/acidcats/tumblrlikes/ui/screens/login_screen/LoginFragment.java
package nl.acidcats.tumblrlikes.ui.screens.login_screen;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.text.InputFilter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.TextView;

import javax.inject.Inject;

import butterknife.BindView;
import nl.acidcats.tumblrlikes.R;
import nl.acidcats.tumblrlikes.di.AppComponent;
import nl.acidcats.tumblrlikes.ui.screens.base.BaseFragment;
import nl.acidcats.tumblrlikes.util.TextWatcherAdapter;

/**
 * Created by stephan on 13/04/2017.
 */

public class LoginFragment extends BaseFragment implements LoginScreenContract.View {
    private static final String TAG = LoginFragment.class.getSimpleName();

    @Inject
    LoginScreenContract.Presenter _presenter;

    @BindView(R.id.input_password)
    EditText _passwordInput;
    @BindView(R.id.tv_pincode_header)
    TextView _header;
    @BindView(R.id.btn_skip)
    View _skipButton;
    @BindView(R.id.tv_pincode_no_match)
    TextView _pincodeNoMatchText;

    private TextWatcherAdapter _textWatcher;

    public static LoginFragment newInstance(LoginScreenContract.Mode mode) {
        Bundle args = new Bundle();
        args.putString(LoginScreenContract.Companion.KEY_MODE, mode.name());

        LoginFragment fragment = new LoginFragment();
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    protected void injectFrom(AppComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        _presenter.restoreState(savedInstanceState, getArguments());
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        _presenter.saveState(outState);

        super.onSaveInstanceState(outState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_login, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        _presenter.setView(this);
        _presenter.onViewCreated();

        _passwordInput.setFilters(new InputFilter[]{new InputFilter.LengthFilter(LoginScreenContract.Companion.PINCODE_LENGTH)});

        _textWatcher = new TextWatcherAdapter() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                _presenter.onPincodeInputChanged(s.toString());
            }
        };
        _passwordInput.addTextChangedListener(_textWatcher);

        _skipButton.setOnClickListener(v -> _presenter.skipLogin());
    }

    @Override
    public void setHeaderTextId(@StringRes int headerTextId) {
        _header.setText(headerTextId);
    }

    @Override
    public void setPincodeDoesntMatchViewVisible(boolean isVisible) {
        _pincodeNoMatchText.setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setSkipButtonVisible(boolean isVisible) {
        _skipButton.setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void clearPasswordInput() {
        _passwordInput.setText("");
    }

    @Override
    public void onResume() {
        super.onResume();

        if (getActivity() != null) {
            getActivity().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        }
    }

    @Override
    public void onDestroyView() {
        _passwordInput.removeTextChangedListener(_textWatcher);
        _skipButton.setOnClickListener(null);

        _presenter.onDestroyView();

        super.onDestroyView();
    }
}

File: app/src/main/java/nl/acidcats/tumblrlikes/ui/screens/login_screen/LoginFragment.kt
package nl.acidcats.tumblrlikes.ui.screens.login_screen

import android.os.Bundle
import android.text.InputFilter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView
import butterknife.BindView
import nl.acidcats.tumblrlikes.R
import nl.acidcats.tumblrlikes.di.AppComponent
import nl.acidcats.tumblrlikes.ui.screens.base.BaseFragment
import nl.acidcats.tumblrlikes.util.TextWatcherAdapter
import javax.inject.Inject

/**
 * Created on 30/10/2018.
 */
class LoginFragment : BaseFragment(), LoginScreenContract.View {

    @Inject
    lateinit var presenter: LoginScreenContract.Presenter

    @BindView(R.id.input_password)
    lateinit var passwordInput: EditText
    @BindView(R.id.tv_pincode_header)
    lateinit var header: TextView
    @BindView(R.id.btn_skip)
    lateinit var skipButton: View
    @BindView(R.id.tv_pincode_no_match)
    lateinit var pincodeNoMatchText: TextView

    private lateinit var textWatcher: TextWatcherAdapter

    companion object {
        fun newInstance(mode: LoginScreenContract.Mode): LoginFragment {
            val args = Bundle()
            args.putString(LoginScreenContract.KEY_MODE, mode.name)

            val fragment = LoginFragment()
            fragment.arguments = args

            return fragment
        }
    }

    override fun injectFrom(appComponent: AppComponent) = appComponent.inject(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        presenter.restoreState(savedInstanceState, arguments)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        presenter.saveState(outState)

        super.onSaveInstanceState(outState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? = inflater.inflate(R.layout.fragment_login, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        presenter.setView(this)
        presenter.onViewCreated()

        passwordInput.filters = arrayOf(InputFilter.LengthFilter(LoginScreenContract.PINCODE_LENGTH))

        textWatcher = object : TextWatcherAdapter() {
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val pincode = s?.toString() ?: ""
                presenter.onPincodeInputChanged(pincode)
            }
        }
        passwordInput.addTextChangedListener(textWatcher)

        skipButton.setOnClickListener { presenter.skipLogin() }
    }

    override fun onResume() {
        super.onResume()


    }

    override fun setPincodeDoesntMatchViewVisible(isVisible: Boolean) {
        pincodeNoMatchText.visibility = if (isVisible) View.VISIBLE else View.GONE
    }

    override fun setSkipButtonVisible(isVisible: Boolean) {
        skipButton.visibility = if (isVisible) View.VISIBLE else View.GONE
    }

    override fun setHeaderTextId(headerTextId: Int) {
        header.text = getString(headerTextId)
    }

    override fun clearPasswordInput() {
        passwordInput.setText("")
    }
}